### PR TITLE
Egao/enrollment card

### PR DESCRIPTION
**Pull request by edwardg8**

### CODE REVIEW

**Comment by edwardg8** [comment]
Will have deeper clean up on on enrollment card once this one gets merged since it's already having too much changes in this PR already. It's involving changing to class-based, remove all the event listeners and got rid of dead functions.

**Comment by edwardg8** [comment]
There are some performance issues with this new change, when hooked up in LMS, loading cards after changing tab is so much slower, will investigate why.

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 1]
```diff
@@ -35,417 +31,407 @@ import 'd2l-card/d2l-card.js';
...
-$_documentContainer.innerHTML = `<dom-module id="d2l-enrollment-card">
-	<template strip-whitespace="">
-		<style>
-			:host {
-				display: block;
-				position: relative;
-			}
+import { html, PolymerElement } from '@polymer/polymer/polymer-element.js';
+import { EntityMixin } from 'siren-sdk/src/mixin/entity-mixin.js';
+import { EnrollmentEntity } from '../../EnrollmentEntity.js';
```
Comment (ciss1995) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: This has moved to `siren-sdk` right?

**Comment by edwardg8** [comment]
The performance issue and pinning issue is fixed, it's ready for review now

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 1 above)
Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: yep, and the other entity too, will move after this PR gets merged.

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 2]
```diff
@@ -35,417 +31,407 @@ import 'd2l-card/d2l-card.js';
...
+	constructor() {
+		super();
+		this._setEntityType(EnrollmentEntity);
+	}
 
-			d2l-card[disabled] d2l-enrollment-updates {
-				display: none;
-			}
+	static get template() {
+		return html`
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: Taking the assumption styles have not been changed.

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 3]
```diff
@@ -35,417 +31,407 @@ import 'd2l-card/d2l-card.js';
...
+				></d2l-enrollment-updates>
+			</d2l-card>
+		`;
+	}
+
+	static get is() { return 'd2l-enrollment-card'; }
+
+	static get properties() {
+		return {
+			href: {
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: We can remove this now

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 4]
```diff
@@ -35,417 +31,407 @@ import 'd2l-card/d2l-card.js';
...
+			showUnreadDiscussionMessages: {
+				type: Boolean,
+				value: false
+			},
+			showUnreadDropboxSubmissions: {
+				type: Boolean,
+				value: false
+			},
+
+			_enrollment: {
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: This is probably just a copy of the `_entity` and not even used.

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 5]
```diff
@@ -35,417 +31,407 @@ import 'd2l-card/d2l-card.js';
...
+				type: Object,
+				value: function() { return {}; }
+			},
+			_selfHref: String,
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: this should always been the same has `_href` ? Unless this is a fix for card swapping.

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 6]
```diff
@@ -35,417 +31,407 @@ import 'd2l-card/d2l-card.js';
...
+			_imageLoading: {
+				type: Boolean,
+				value: false
+			},
+			_imageLoadingProgress: {
+				type: Boolean,
+				value: false
+			},
+			_load: Boolean,
+			_organization: Object,
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: We don't need this anymore.

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 7]
```diff
@@ -35,417 +31,407 @@ import 'd2l-card/d2l-card.js';
...
+					padding: 0.2rem 0.5rem 0.15rem 0.5rem;
+					filter: grayscale(0);
+					opacity: 1;
+				}
+			</style>
+
+			<d2l-card disabled$="[[disabled]]" href="[[_organizationHomepageUrl]]" text="[[_accessibilityDataToString(_accessibilityData)]]">
+				<div slot="header" aria-hidden="true">
+					<div class="d2l-enrollment-card-image-container">
+						<d2l-course-image image="[[_image]]" sizes="[[_tileSizes]]" type="tile">
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: Next PR switch to `d2l-organization-image`

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 8]
```diff
@@ -501,38 +496,24 @@ Polymer({
 
 		this._organizationUrl = organization.getLinkByRel('self').href;
 
-		return this._entityStoreFetch(this._organizationUrl)
-			.then(this._handleOrganizationResponse.bind(this))
+		if (!this._entity) {
+			return;
+		}
+
+		return this._entity.onOrganizationChange(this.onOrganizationChange.bind(this))
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: onOrganizationChange doesn't return a promise. Maybe it should?

File: components/d2l-enrollment-card/d2l-enrollment-card.js [hunk 9]
```diff
@@ -594,91 +576,157 @@ Polymer({
...
+			this._setUserActivityUsageAccessible(dateTextAndStatus && dateTextAndStatus.dateText);
+			if (!userActivityUsage.isAttended()) {
+				this._handleEnrollmentNew();
+			}
+		});
+
+		enrollment.onOrganizationChange(this.onOrganizationChange.bind(this));
+	}
+
+	onOrganizationChange(org) {
```
Comment (m6jones) [inline components/d2l-enrollment-card/d2l-enrollment-card.js]: This needs to be `enrollment.onOrganizationChange` -- Hope this isn't the cause of the performance increase

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 2 above)
Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: Template is not touched, yea

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 4 above)
Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: this._enrollment is still referred in quite a few places in the event, in next a few PR, I'm trying to have it removed. https://github.com/BrightspaceHypermediaComponents/enrollments/blob/master/components/d2l-enrollment-card/d2l-enrollment-card.js#L780

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 5 above)
Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: I think you are right, will remove this.

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 6 above)
Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: this is still used in many places too, on my target to have it removed later.

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 9 above)
Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: this is the onChange function passed into `enrollment.onOrganizationChange`

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 8 above)
Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: oh, can we make it support it, merely for this guy I think https://github.com/BrightspaceHypermediaComponents/enrollments/blob/master/components/d2l-enrollment-card/d2l-enrollment-card.js#L505

Comment (m6jones) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: I think the then and catch should be within the `onOrganizationChange`

Comment (edwardg8) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: that's quick workaround, i'll move the code down there.

File: components/d2l-enrollment-card/d2l-enrollment-card.js (hunk 9 above)
Comment (m6jones) [reply components/d2l-enrollment-card/d2l-enrollment-card.js]: I miss read this :( I saw it in the method not as a new method. Oops

**Pull request merged**